src/rabbit_dead_letter.erl

```diff
@@ -84,36 +84,85 @@ x_death_event_key(Info, Key, KeyType) ->
         {value, {Key, KeyType, Val}} -> Val
     end.
 
+maybe_append_to_event_group(Info, _Queue, _Reason, []) ->
+    [Info];
+maybe_append_to_event_group(Info, Queue, Reason, Acc) ->
+    case lists:any(queue_and_reason_matcher(Queue, Reason), Acc) of
+        true  -> Acc;
+        false -> [Info | Acc]
+    end.
+
+group_by_queue_and_reason(Tables) ->
+    Infos = [Info || {table, Info} <- Tables],
+    Grouped =
+        lists:foldl(fun (Info, Acc) ->
+            Q = x_death_event_key(Info, <<"queue">>, longstr),
+            R = x_death_event_key(Info, <<"reason">>, longstr),
+            Matcher = queue_and_reason_matcher(Q, R),
+            {Matches, _} = lists:partition(Matcher, Infos),
```

Comment:
You could sort the list inside `group_by_queue_and_reason` before calling the fold. I think currently by calling `lists:any` all the time the complexity of this function could go quite up since you are always traversing the Accumulator for every iteration of the foldl. Sorting the input before calling foldl removes that